### PR TITLE
Use storage plugin for pos

### PR DESCRIPTION
**Pull request by cosmo0920**
I've confirmed that this patches work with the following config and Fluentd v0.14.12:

```log
<source>
  @type windows_eventlog
  channels ["Application", "Windows PowerShell"]
  read_interval 1
  tag winevt.raw
  from_encoding SHIFT_JIS
  encoding UTF-8
  @id windows_eventlog
</source>
<match **>
  @type stdout
</match>
<system>
  root_dir .\tmp
</system>
```

```log
2017-01-30 18:49:57 +0900 [info]: #0 starting fluentd worker pid=10324 ppid=7360 worker=0
2017-01-30 18:49:57 +0900 [info]: #0 fluentd worker is now running worker=0
2017-01-30 18:49:57.371701000 +0900 fluent.info: {"worker":0,"message":"fluentd worker is now running worker=0"}
2017-01-30 18:50:09.378418000 +0900 winevt.raw: {"channel":"windows powershell","record_number":"1599","time_generated":"2017-01-30 18:50:08 +0900","time_written":"2017-01-30 18:50:08 +0900","event_id":"600","event_type":"information","event_category":"6","source_name":"PowerShell","computer_name":"LAPTOP-2KFM0IP4","user":"","description":"プロバイダー \"Registry\" は Startedです。\r\n\r\n詳細: \r\n\tProviderName=Registry\r\n\tNewProviderState=Started\r\n\r\n\tSequenceNumber=1\r\n\r\n\tHostName=ConsoleHost\r\n\tHostVersion=5.1.14393.693\r\n\tHostId=f53c097c-4b3d-40f4-a168-381378b4e181\r\n\tHostApplication=C:\\WINDOWS\\System32\\WindowsPowerShell\\v1.0\\Powershell.exe -NoExit -ExecutionPolicy Unrestricted -File C:\\Users\\hhatake\\AppData\\Local\\GitHub\\PoshGit_a2be688889e1b24632e83adccd9b2a44b91d655b\\profile.example.ps1\r\n\tEngineVersion=\r\n\tRunspaceId=\r\n\tPipelineId=\r\n\tCommandName=\r\n\tCommandType=\r\n\tScriptName=\r\n\tCommandPath=\r\n\tCommandLine=\r\n"}
2017-01-30 18:50:09.381426000 +0900 winevt.raw: {"channel":"windows powershell","record_number":"1600","time_generated":"2017-01-30 18:50:08 +0900","time_written":"2017-01-30 18:50:08 +0900","event_id":"600","event_type":"information","event_category":"6","source_name":"PowerShell","computer_name":"LAPTOP-2KFM0IP4","user":"","description":"プロバイダー \"Alias\" は Startedです。\r\n\r\n詳細: \r\n\tProviderName=Alias\r\n\tNewProviderState=Started\r\n\r\n\tSequenceNumber=3\r\n\r\n\tHostName=ConsoleHost\r\n\tHostVersion=5.1.14393.693\r\n\tHostId=f53c097c-4b3d-40f4-a168-381378b4e181\r\n\tHostApplication=C:\\WINDOWS\\System32\\WindowsPowerShell\\v1.0\\Powershell.exe -NoExit -ExecutionPolicy Unrestricted -File C:\\Users\\hhatake\\AppData\\Local\\GitHub\\PoshGit_a2be688889e1b24632e83adccd9b2a44b91d655b\\profile.example.ps1\r\n\tEngineVersion=\r\n\tRunspaceId=\r\n\tPipelineId=\r\n\tCommandName=\r\n\tCommandType=\r\n\tScriptName=\r\n\tCommandPath=\r\n\tCommandLine=\r\n"}
2017-01-30 18:50:09.382428000 +0900 winevt.raw: {"channel":"windows powershell","record_number":"1601","time_generated":"2017-01-30 18:50:08 +0900","time_written":"2017-01-30 18:50:08 +0900","event_id":"600","event_type":"information","event_category":"6","source_name":"PowerShell","computer_name":"LAPTOP-2KFM0IP4","user":"","description":"プロバイダー \"Environment\" は Startedです。\r\n\r\n詳細: \r\n\tProviderName=Environment\r\n\tNewProviderState=Started\r\n\r\n\tSequenceNumber=5\r\n\r\n\tHostName=ConsoleHost\r\n\tHostVersion=5.1.14393.693\r\n\tHostId=f53c097c-4b3d-40f4-a168-381378b4e181\r\n\tHostApplication=C:\\WINDOWS\\System32\\WindowsPowerShell\\v1.0\\Powershell.exe -NoExit -ExecutionPolicy Unrestricted -File C:\\Users\\hhatake\\AppData\\Local\\GitHub\\PoshGit_a2be688889e1b24632e83adccd9b2a44b91d655b\\profile.example.ps1\r\n\tEngineVersion=\r\n\tRunspaceId=\r\n\tPipelineId=\r\n\tCommandName=\r\n\tCommandType=\r\n\tScriptName=\r\n\tCommandPath=\r\n\tCommandLine=\r\n"}
2017-01-30 18:50:09.382428000 +0900 winevt.raw: {"channel":"windows powershell","record_number":"1602","time_generated":"2017-01-30 18:50:08 +0900","time_written":"2017-01-30 18:50:08 +0900","event_id":"600","event_type":"information","event_category":"6","source_name":"PowerShell","computer_name":"LAPTOP-2KFM0IP4","user":"","description":"プロバイダー \"FileSystem\" は Startedです。\r\n\r\n詳細: \r\n\tProviderName=FileSystem\r\n\tNewProviderState=Started\r\n\r\n\tSequenceNumber=7\r\n\r\n\tHostName=ConsoleHost\r\n\tHostVersion=5.1.14393.693\r\n\tHostId=f53c097c-4b3d-40f4-a168-381378b4e181\r\n\tHostApplication=C:\\WINDOWS\\System32\\WindowsPowerShell\\v1.0\\Powershell.exe -NoExit -ExecutionPolicy Unrestricted -File C:\\Users\\hhatake\\AppData\\Local\\GitHub\\PoshGit_a2be688889e1b24632e83adccd9b2a44b91d655b\\profile.example.ps1\r\n\tEngineVersion=\r\n\tRunspaceId=\r\n\tPipelineId=\r\n\tCommandName=\r\n\tCommandType=\r\n\tScriptName=\r\n\tCommandPath=\r\n\tCommandLine=\r\n"}
2017-01-30 18:50:09.382428000 +0900 winevt.raw: {"channel":"windows powershell","record_number":"1603","time_generated":"2017-01-30 18:50:08 +0900","time_written":"2017-01-30 18:50:08 +0900","event_id":"600","event_type":"information","event_category":"6","source_name":"PowerShell","computer_name":"LAPTOP-2KFM0IP4","user":"","description":"プロバイダー \"Function\" は Startedです。\r\n\r\n詳細: \r\n\tProviderName=Function\r\n\tNewProviderState=Started\r\n\r\n\tSequenceNumber=9\r\n\r\n\tHostName=ConsoleHost\r\n\tHostVersion=5.1.14393.693\r\n\tHostId=f53c097c-4b3d-40f4-a168-381378b4e181\r\n\tHostApplication=C:\\WINDOWS\\System32\\WindowsPowerShell\\v1.0\\Powershell.exe -NoExit -ExecutionPolicy Unrestricted -File C:\\Users\\hhatake\\AppData\\Local\\GitHub\\PoshGit_a2be688889e1b24632e83adccd9b2a44b91d655b\\profile.example.ps1\r\n\tEngineVersion=\r\n\tRunspaceId=\r\n\tPipelineId=\r\n\tCommandName=\r\n\tCommandType=\r\n\tScriptName=\r\n\tCommandPath=\r\n\tCommandLine=\r\n"}
2017-01-30 18:50:09.382428000 +0900 winevt.raw: {"channel":"windows powershell","record_number":"1604","time_generated":"2017-01-30 18:50:08 +0900","time_written":"2017-01-30 18:50:08 +0900","event_id":"600","event_type":"information","event_category":"6","source_name":"PowerShell","computer_name":"LAPTOP-2KFM0IP4","user":"","description":"プロバイダー \"Variable\" は Startedです。\r\n\r\n詳細: \r\n\tProviderName=Variable\r\n\tNewProviderState=Started\r\n\r\n\tSequenceNumber=11\r\n\r\n\tHostName=ConsoleHost\r\n\tHostVersion=5.1.14393.693\r\n\tHostId=f53c097c-4b3d-40f4-a168-381378b4e181\r\n\tHostApplication=C:\\WINDOWS\\System32\\WindowsPowerShell\\v1.0\\Powershell.exe -NoExit -ExecutionPolicy Unrestricted -File C:\\Users\\hhatake\\AppData\\Local\\GitHub\\PoshGit_a2be688889e1b24632e83adccd9b2a44b91d655b\\profile.example.ps1\r\n\tEngineVersion=\r\n\tRunspaceId=\r\n\tPipelineId=\r\n\tCommandName=\r\n\tCommandType=\r\n\tScriptName=\r\n\tCommandPath=\r\n\tCommandLine=\r\n"}
2017-01-30 18:50:09.383431000 +0900 winevt.raw: {"channel":"windows powershell","record_number":"1605","time_generated":"2017-01-30 18:50:08 +0900","time_written":"2017-01-30 18:50:08 +0900","event_id":"400","event_type":"information","event_category":"4","source_name":"PowerShell","computer_name":"LAPTOP-2KFM0IP4","user":"","description":"エンジンの状態が None から Available に変更されました。\r\n\r\n詳細: \r\n\tNewEngineState=Available\r\n\tPreviousEngineState=None\r\n\r\n\tSequenceNumber=13\r\n\r\n\tHostName=ConsoleHost\r\n\tHostVersion=5.1.14393.693\r\n\tHostId=f53c097c-4b3d-40f4-a168-381378b4e181\r\n\tHostApplication=C:\\WINDOWS\\System32\\WindowsPowerShell\\v1.0\\Powershell.exe -NoExit -ExecutionPolicy Unrestricted -File C:\\Users\\hhatake\\AppData\\Local\\GitHub\\PoshGit_a2be688889e1b24632e83adccd9b2a44b91d655b\\profile.example.ps1\r\n\tEngineVersion=5.1.14393.693\r\n\tRunspaceId=f9b4194c-d7e5-411c-808d-44c885364597\r\n\tPipelineId=\r\n\tCommandName=\r\n\tCommandType=\r\n\tScriptName=\r\n\tCommandPath=\r\n\tCommandLine=\r\n"}
```

### CODE REVIEW

File: lib/fluent/plugin/in_windows_eventlog.rb [hunk 1]
```diff
@@ -21,12 +22,14 @@ class WindowsEventLogInput < Input
...
     config_param :read_interval, :time, default: 2
-    config_param :pos_file, :string, default: nil
+    config_param :pos_file, :string, default: nil,
+                 obsoleted: "This section is not used anymore. Use 'store_pos' instead."
     config_param :channels, :array, default: ['Application']
     config_param :keys, :string, default: []
     config_param :read_from_head, :bool, default: false
     config_param :from_encoding, :string, default: nil
     config_param :encoding, :string, default: nil
+    config_param :store_pos, :bool, default: false
```
Comment (tagomoris) [inline lib/fluent/plugin/in_windows_eventlog.rb]: `pos_file` is obsoleted, so this plugin doesn't work without specifying `store_pos true`.
So, this parameter doesn't have to exist (we can remove this parameter), and we suppose that this plugin always uses storage plugin to store positions.

File: lib/fluent/plugin/in_windows_eventlog.rb [hunk 2]
```diff
@@ -55,6 +59,20 @@ def configure(conf)
                           else
                             method(:no_encode_record)
                           end
+      if @store_pos
+        unless system_config.root_dir
```
Comment (tagomoris) [inline lib/fluent/plugin/in_windows_eventlog.rb]: If `persistent true`, storage plugin should raise configuration errors if any paths are unavailable.
So this check is not needed, and moreover, this check prohibits users to specify `<storage> path </storage>` configuration sections manually.

File: lib/fluent/plugin/in_windows_eventlog.rb [hunk 3]
```diff
@@ -55,6 +59,20 @@ def configure(conf)
...
+          raise Fluent::ConfigError, "windows_eventlog: 'root_dir' in <system> parameter must be specified with 'store_pos' parameter."
+        end
+        @chs.map {|ch|
```
Comment (tagomoris) [inline lib/fluent/plugin/in_windows_eventlog.rb]: `@chs.each` is better.

File: lib/fluent/plugin/in_windows_eventlog.rb [hunk 4]
```diff
@@ -55,6 +59,20 @@ def configure(conf)
...
+          config = Fluent::Config::Element.new('storage',
+                                               "in_windows_eventlog_pos_#{ch.gsub(' ', '_')}", {
```
Comment (tagomoris) [inline lib/fluent/plugin/in_windows_eventlog.rb]: Storage usage should be short and simple: it should be unique per owner plugins, not globally.
It looks enough to use channel name as usage directly.

Comment (tagomoris) [inline lib/fluent/plugin/in_windows_eventlog.rb]: `ch.gsub(' ', '_')` is not enough to escape paths.
`ch.gsub(/[^0-9a-zA-Z]/, ' ')`? (tests required)

File: lib/fluent/plugin/in_windows_eventlog.rb [hunk 5]
```diff
@@ -55,6 +59,20 @@ def configure(conf)
...
+                                                 "@type" => "local",
+                                                 "persistent" => true,
+                                               }, [])
+          @storages[ch] = storage_create(usage: "in_windows_eventlog_pos_#{ch.gsub(' ', '_')}", conf: config,
```
Comment (tagomoris) [inline lib/fluent/plugin/in_windows_eventlog.rb]: It's not good enough to use this style method call for `storage_create` (it prevents users to write their own configuration section as `<storage channel>`).
I think it works well to use just `usage` with `config_section` to specify default parameter (for `@type` and `persistent`).

```ruby
config_section :storage do
  config_set_default :@type, DEFAULT_STORAGE_TYPE
  config_set_default :persistent, true
end

def configure(conf)
  # ...
  @chs.each do |ch|
    escaped_ch = ch.gsub(/[^0-9a-zA-Z]/, '_')
    @storages[ch] = storage_create(usage: escaped_usage)
  end
end
```

File: lib/fluent/plugin/in_windows_eventlog.rb [hunk 6]
```diff
@@ -92,17 +110,18 @@ def no_encode_record(record)
 
     def start
       super
-      if @pos_file
-        @pf_file = File.open(@pos_file, File::RDWR|File::CREAT|File::BINARY)
-        @pf_file.sync = true
-        @pf = PositionFile.parse(@pf_file)
+      unless @storages.empty?
+        @pf = {}
+        @storages.each {|storage|
```
Comment (tagomoris) [inline lib/fluent/plugin/in_windows_eventlog.rb]: What this code is for? Storage plugins loads data from files at `#start` (via storage plugin helper) in default.

File: lib/fluent/plugin/in_windows_eventlog.rb [hunk 7]
```diff
@@ -92,229 +101,77 @@ def no_encode_record(record)
...
-        end
-        @file.pos = @last_pos
-        @file.write ch
-        @file.write "\t"
-        seek = @file.pos
-        @file.write "00000000\t00000000\n"
-        @last_pos = @file.pos
-        @map[ch] = FilePositionEntry.new(@file, seek)
-      end
+      numlines = current_end - old_end
```
Comment (repeatedly) [inline lib/fluent/plugin/in_windows_eventlog.rb]: numlines is not used. This value is needed?

Comment (cosmo0920) [inline lib/fluent/plugin/in_windows_eventlog.rb]: Oops, it isn't needed.

**Comment by repeatedly** [comment]
Before merging, could you update README for mentiong `<storage>`?

**Pull request merged**